fuzzing github.com/andybalholm/cascadia crashes versifier

When I try to fuzz the CSS selector parsing code in github.com/andybalholm/cascadia, I get the following panic:

panic: runtime error: slice bounds out of range

goroutine 28 [running]:
github.com/dvyukov/go-fuzz/go-fuzz/versifier.structureKeyValue(0xc820328180, 0x2, 0x2, 0x0, 0x0, 0x0)
	/Users/andy/src/github.com/dvyukov/go-fuzz/go-fuzz/versifier/versifier.go:696 +0x489
github.com/dvyukov/go-fuzz/go-fuzz/versifier.structure(0xc820328180, 0x2, 0x2, 0x0, 0x0, 0x0)
	/Users/andy/src/github.com/dvyukov/go-fuzz/go-fuzz/versifier/versifier.go:575 +0xdd
github.com/dvyukov/go-fuzz/go-fuzz/versifier.BuildVerse(0xc82027a1b0, 0xc82020a960, 0x6, 0x6, 0x10000)
	/Users/andy/src/github.com/dvyukov/go-fuzz/go-fuzz/versifier/versifier.go:55 +0x1b9
main.(*Hub).loop(0xc820224100)
	/Users/andy/src/github.com/dvyukov/go-fuzz/go-fuzz/hub.go:236 +0x1c2c
created by main.newHub
	/Users/andy/src/github.com/dvyukov/go-fuzz/go-fuzz/hub.go:115 +0xe83

--- a/go-fuzz/versifier/versifier.go
+++ b/go-fuzz/versifier/versifier.go
@@ -571,8 +571,8 @@ func tokenize(data []byte) []Node {
 func structure(nn []Node) []Node {
 	nn = extractNumbers(nn)
 	nn = structureBrackets(nn)
-	nn = structureLists(nn)
 	nn = structureKeyValue(nn)
+	nn = structureLists(nn)
 	nn = structureLines(nn)
 	return nn
 }

--- a/go-fuzz/versifier/versifier.go
+++ b/go-fuzz/versifier/versifier.go
@@ -678,19 +678,16 @@ func structureKeyValue(nn []Node) (res []Node) {
 		if !ok {
 			continue
 		}
-		if delims[ctrl.ch] {
+		if delims[ctrl.ch] &&
+			!(i == 0 || i == len(nn)-1) {
 			var key, value *AlphaNumNode
-			if i > 0 {
-				key, ok = nn[i-1].(*AlphaNumNode)
-				if !ok {
-					continue
-				}
+			key, ok = nn[i-1].(*AlphaNumNode)
+			if !ok {
+				continue
 			}
-			if i < len(nn)-1 {
-				value, ok = nn[i+1].(*AlphaNumNode)
-				if !ok {
-					continue
-				}
+			value, ok = nn[i+1].(*AlphaNumNode)
+			if !ok {
+				continue
 			}
 			nn[i+1] = &KeyValNode{ctrl.ch, key, value}
 			copy(nn[i-1:], nn[i+1:])

--- a/go-fuzz/versifier/versifier.go
+++ b/go-fuzz/versifier/versifier.go
@@ -671,6 +671,11 @@ func extractNumbers(nn []Node) []Node {
 func structureKeyValue(nn []Node) (res []Node) {
 	// TODO: extract numeric key-value pairs
 	delims := map[rune]bool{'=': true, ':': true}
+	for _, n := range nn {
+		if brk, ok := n.(*BracketNode); ok {
+			brk.b.nodes = structureKeyValue(brk.b.nodes)
+		}
+	}
 
 	for i := 0; i < len(nn); i++ {
 		n := nn[i]

--- a/go-fuzz/versifier/versifier_test.go
+++ b/go-fuzz/versifier/versifier_test.go
@@ -27,5 +27,6 @@ func TestBracket(t *testing.T) {
 }
 
 func TestKeyValue(t *testing.T) {
-	dump(`a=1 a=b   2  (aa=bb) a bb:cc`)
+	dump(`a=1 a=b   2  (aa=bb) a bb:cc:dd,a=b,c=d,e=f`)
+	dump(`:a`)
 }

